### PR TITLE
Add the API to set the session configuration when load the model in the JAVA

### PR DESCRIPTION
**Pull request by raintung**
See the details in the issue:
[https://github.com/tensorflow/tensorflow/issues/18143]<https://github.com/tensorflow/tensorflow/issues/18143>

### CODE REVIEW

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java [hunk 1]
```diff
@@ -36,9 +36,22 @@
...
   }
-
+   
+    /**
+   * Load a saved model from an export directory. The model that is being loaded should be created
+   * using the <a href="https://www.tensorflow.org/api_docs/python/tf/saved_model">Saved Model
+   * API</a>.
+   *
+   * @param exportDir the directory path containing a saved model.
+   * @param tags the tags identifying the specific metagraphdef to load.
```
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Please add appropriate documentation for `runOptions` and `sessionOptions`

File: tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java [hunk 2]
```diff
@@ -36,9 +36,22 @@
...
+   * @return a bundle containing the graph and associated session.
+   */
+  public static SavedModelBundle load(String exportDir,byte[] runOptions, byte[] sessionOptions, String... tags) {
```
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: Mind formatting the file as per [`google-java-format`](https://github.com/google/google-java-format)?

Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/SavedModelBundle.java]: It appears that `sessionOptions` here is just the serialized `ConfigProto` message, not something equivalent to [`TF_SessionOptions`](https://github.com/tensorflow/tensorflow/blob/r1.7/tensorflow/c/c_api.h#L304) which has both a [`ConfigProto`](https://github.com/tensorflow/tensorflow/blob/r1.7/tensorflow/c/c_api.h#L322) and a ["target" address](https://github.com/tensorflow/tensorflow/blob/r1.7/tensorflow/c/c_api.h#L315)  associated with it.

These are a bunch of options, so I'm wondering that instead of having a single function that takes multiple arguments, would it make sense to use the builder pattern for this? So calling code would look something like so:

```java
SavedModelBundle model = SavedModelBundle.loader("/tmp/exportdir")
    .withConfig(configProto)
    .withRunOptions(runOptions)
    .addTag("serve")
    .load();
```

Using something like:

```java
public class SavedModelBundle {
  public final class Loader {
    public Loader withRunOptions(byte[] options) {
      this.runOptions = options;
      return this;
    }
    public Loader withSessionConfig(byte[] config) {
     this.config = config;
     return this;
    }
    public Loader addTag(String tag) ...

    public SavedModelBundle load() { ... }

    private Loader(String exportDir) ...
    // etc
 
   public static Loader loader(String exportDir) { return new Loader(exportDir);  }
   ...
}
```

The private JNI function though could of course just be a flat function that takes all the arguments.

Thoughts?

File: tensorflow/java/src/main/native/saved_model_bundle_jni.cc [hunk 3]
```diff
@@ -20,14 +20,36 @@ limitations under the License.
...
-    jbyteArray run_options) {
+    jbyteArray run_options, jbyteArray session_options) {
   TF_Status* status = TF_NewStatus();
   jobject bundle = nullptr;
 
   // allocate parameters for TF_LoadSessionFromSavedModel
   TF_SessionOptions* opts = TF_NewSessionOptions();
+
+
+  TF_Buffer* csession_options = nullptr;
```
Comment (asimshankar) [inline tensorflow/java/src/main/native/saved_model_bundle_jni.cc]: The formatting here seems to be different from the conventions used in the rest of the file (e.g., two-spaces instead of 4). Could you remove that inconsistency?

(In general, the native code should [follow the Google style guide](https://google.github.io/styleguide/cppguide.html))

File: tensorflow/java/src/main/native/saved_model_bundle_jni.cc [hunk 4]
```diff
@@ -20,14 +20,36 @@ limitations under the License.
...
+		  jbyte* sconfigbyte = env->GetByteArrayElements(session_options, nullptr);
+		  csession_options =
+		           TF_NewBufferFromString(static_cast<void*>(sconfigbyte), sz);
+		  env->ReleaseByteArrayElements(session_options, sconfigbyte, JNI_ABORT);
+	      TF_SetConfig(opts, csession_options->data,
+					  	  	  	  	  csession_options->length, status);
+	      if (!throwExceptionIfNotOK(env, status)) {
+	    	   	   TF_DeleteSessionOptions(opts);
+	           TF_DeleteStatus(status);
+	           TF_DeleteBuffer(csession_options);
```
Comment (asimshankar) [inline tensorflow/java/src/main/native/saved_model_bundle_jni.cc]: You should be able to unconditionally call `TF_DeleteBuffer` after calling `TF_SetConfig`.

File: tensorflow/java/src/main/native/session_jni.cc [hunk 5]
```diff
@@ -113,26 +113,43 @@ JNIEXPORT jlong JNICALL Java_org_tensorflow_Session_allocate2(
     throwException(env, kNullPointerException, "Graph has been close()d");
     return 0;
   }
-  TF_Graph* graph = reinterpret_cast<TF_Graph*>(graph_handle);
+
```
Comment (asimshankar) [inline tensorflow/java/src/main/native/session_jni.cc]: Why the changes to this file?

**Review by asimshankar — CHANGES_REQUESTED** [review]
Thanks for the contribution!

**Comment by asimshankar** [comment]
Also, it seems this PR is aiming for a merge into the 1.6 release branch, I don't think that's what you want. The PR should be for the master branch, no?

**Review by raintung — COMMENTED** [review]
I change the code as requirement

File: tensorflow/java/src/main/native/saved_model_bundle_jni.cc (hunk 4 above)
Comment (asimshankar) [reply tensorflow/java/src/main/native/saved_model_bundle_jni.cc]: I apologize, I didn't quite follow your comment.
I was suggesting that instead of calling `TF_DeleteBuffer(csession_options)` in two places (once inside the `if (!throwExceptionIfNotOK(...` block and once in line 87, we could call it immediately after  invoking `TF_SetConfig`.

File: tensorflow/java/src/main/native/session_jni.cc (hunk 5 above)
Comment (asimshankar) [reply tensorflow/java/src/main/native/session_jni.cc]: Right, but the change here seems more than just plugging those leaks. Why bother with `TF_NewBufferFromString`? Can't we reduce the change in this file to 2 lines:

```c++
if (!throwExcpetionIfNotOK(env, status)) {
  env->ReleaseByteArrayElemenst(config, cconfig, JNI_ABORT);
  TF_DeleteSessionOptions(opts);
  TF_DeleteStatus(status);
  return 0;
}
```
?

Comment (raintung) [reply tensorflow/java/src/main/native/session_jni.cc]: Yes, it look simple but I have some reasons as blow.
1.  Keep the code same, please take a look for saved_model_bundle_jni.cc code for run_options logic
`  if (run_options != nullptr) {
    size_t sz = env->GetArrayLength(run_options);
    if (sz > 0) {
      jbyte* run_options_data = env->GetByteArrayElements(run_options, nullptr);
      crun_options =
          TF_NewBufferFromString(static_cast<void*>(run_options_data), sz);
      env->ReleaseByteArrayElements(run_options, run_options_data, JNI_ABORT);
    }
  }`
2. I concern for method TF_SetConfig `options.config.ParseFromArray(proto, proto_len)`. It is protocol buffer logic,  I am not sure it use or copy the proto bytearray into options.config. The last logic will delete options, the proto bytearray should be released in java garbage collect.

**Review by raintung — COMMENTED** [review]
Just common

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 34 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by martinwicke** [comment]
@asimshankar can you take a look?

File: tensorflow/java/src/main/native/session_jni.cc (hunk 5 above)
Comment (asimshankar) [reply tensorflow/java/src/main/native/session_jni.cc]: Thanks, but I'm not sure I follow.

1. `saved_model_bundle_jni.cc` has to use `TF_NewBufferFromString` because the C API function it eventually invokes (`TF_LoadSessionFromSavedModel`) requires a `TF_Buffer*` argument. That is not the case here.

2. `TF_SetConfig` does not hold on to memory of the passed in arguments upon returning, so I don't think you need to be concerned.

**Review by asimshankar — COMMENTED** [review]
Also, please see my [earlier comment](https://github.com/tensorflow/tensorflow/pull/18716/files/f8a7579d35e93d31eca002f43633f41e47d3d15d#issuecomment-383358870), I'm pretty sure you want this PR to merge into the master branch not the r1.6 branch, right?

**Comment by gunan** [comment]
Closing PR to the release branch.

**Comment by asimshankar** [comment]
@raintung - could you recreate the PR for the master branch?
